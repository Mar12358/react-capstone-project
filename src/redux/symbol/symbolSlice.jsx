import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  symbol: [],
  isLoading: false,
  error: undefined,
};

export const getSymbol = (id) => createAsyncThunk(
  'Symbol/getSymbol',
  async (thunkAPI) => {
    try {
      const resp = await axios(`https://api.coincap.io/v2/assets/${id}`);
      console.log(resp.data);
      return resp.data;
    } catch (error) {
      return thunkAPI.rejectWithValue(error);
    }
  },
);

const symbolSlice = createSlice({
  name: 'symbol',
  initialState,
  reducers: {
    /* reserveRocket: (state, action) => {
      const rocketId = action.payload;
      const newState = state;
      newState.assets = state.rockets.map((rocket) => (
        rocket.rocket_id !== rocketId ? rocket : { ...rocket, reserved: true }
      ));
    },
    cancelReserveRocket: (state, action) => {
      const rocketId = action.payload;
      const newState = state;
      newState.rockets = state.rockets.map((rocket) => (
        rocket.rocket_id !== rocketId ? rocket : { ...rocket, reserved: false }
      ));
    }, */
  },
  extraReducers: (builder) => {
    builder
      .addCase(getSymbol.pending, (state) => {
        const pendingState = state;
        pendingState.isLoading = true;
      })
      .addCase(getSymbol.fulfilled, (state, action) => {
        const fulfilledState = state;
        fulfilledState.isLoading = false;
        fulfilledState.symbol = action.payload;
      })
      .addCase(getSymbol.rejected, (state) => {
        const rejectedState = state;
        rejectedState.isLoading = false;
      });
  },
});

// export const { reserveRocket, cancelReserveRocket } = assetsSlice.actions;
export default symbolSlice.reducer;
