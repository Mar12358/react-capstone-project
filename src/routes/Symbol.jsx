import React, { useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { getSymbol } from '../redux/symbol/symbolSlice';
import styles from '../styles/symbol.module.css';

const DetailPage = () => {
  const dispatch = useDispatch();
  const { symbolId } = useParams();
  const { symbol, isLoading, error } = useSelector((state) => state.symbol);
  useEffect(() => {
    // Fetch symbol data using the symbolId
    // Update the symbolData state with the fetched data
    // Example:
    // const fetchData = async () => {
    //   const response = await fetch(`/api/symbols/${symbolId}`);
    //   const data = await response.json();
    //   setSymbolData(data);
    // };
    // fetchData();
    dispatch(getSymbol(symbolId));
  }, [dispatch, symbolId]);

  if (error) return <h2>Something went wrong</h2>;

  if (isLoading) {
    return <div>Loading...</div>;
  }

  return (
    <div>
      <h2>Symbol Detail</h2>
      {/* Display the symbol detail using the symbolData */}
      {/* Example: */}
      {/* <div>Symbol Name: {symbolData.name}</div> */}
      {/* <div>Symbol Price: {symbolData.price}</div> */}
    </div>
  );
};

export default DetailPage;
