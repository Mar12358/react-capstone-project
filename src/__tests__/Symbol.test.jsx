import React from 'react';
import { render, screen } from '@testing-library/react';
import { useParams } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import Symbol from '../routes/Symbol';
import { getSymbol } from '../redux/symbol/symbolSlice';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn(),
}));

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('../redux/symbol/symbolSlice', () => ({
  getSymbol: jest.fn(),
}));

describe('Symbol', () => {
  const mockDispatch = jest.fn();

  beforeEach(() => {
    useParams.mockReturnValue({ id: 'bitcoin' });
    useSelector.mockReturnValue({
      symbol: {
        data: {
          name: 'Bitcoin',
          symbol: 'BTC',
          priceUsd: '50000',
        },
      },
      isLoading: false,
    });
    useDispatch.mockReturnValue(mockDispatch);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('dispatches getSymbol action with correct ID', () => {
    render(<Symbol />);

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(getSymbol).toHaveBeenCalledWith('bitcoin');
  });

  test('renders symbol details correctly', () => {
    render(<Symbol />);

    expect(screen.getByText('name')).toBeInTheDocument();
    expect(screen.getByText('Bitcoin')).toBeInTheDocument();

    expect(screen.getByText('symbol')).toBeInTheDocument();
    expect(screen.getByText('BTC')).toBeInTheDocument();

    expect(screen.getByText('priceUsd')).toBeInTheDocument();
    expect(screen.getByText('50000')).toBeInTheDocument();

    expect(screen.getByText('name')).toMatchSnapshot();
    expect(screen.getByText('symbol')).toMatchSnapshot();
    expect(screen.getByText('priceUsd')).toMatchSnapshot();
  });

  test('renders loading state', () => {
    useSelector.mockReturnValueOnce({ symbol: {}, isLoading: true });

    render(<Symbol />);

    expect(screen.getByText('Loading...')).toBeInTheDocument();
    expect(screen.getByText('Loading...')).toMatchSnapshot();
  });
});
