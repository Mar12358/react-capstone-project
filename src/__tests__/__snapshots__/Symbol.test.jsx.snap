// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Symbol renders loading state 1`] = `
<h2>
  Loading...
</h2>
`;

exports[`Symbol renders symbol details correctly 1`] = `
<p>
  name
</p>
`;

exports[`Symbol renders symbol details correctly 2`] = `
<p>
  symbol
</p>
`;

exports[`Symbol renders symbol details correctly 3`] = `
<p>
  priceUsd
</p>
`;
